### PR TITLE
Allow search table names in mixed cases

### PR DESCRIPTION
**Pull request by coresh**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| Fixed issues | 

#### Summary

Updated pull request

Mysql / MariaDB / Persona allows to use the table name in different cases.

The table names might look like this:

UPPERCASETABLENAME
CamelCaseTableName
lowercasetablename
Titlecasetablename

However, the original implementation of the tablesExist() function searches for existing tables ONLY in lower case:
by using the function:
'strtolower()',

before:
```
$names = array_map('strtolower', (array) $names);

return count($names) === count(array_intersect($names, array_map('strtolower', $this->listTableNames())));
```
after:
```
$tableNames = $this->listTableNames();

if ($lowercase === true) {
    $names      = array_map('strtolower', (array) $names);
    $tableNames = array_map('strtolower', $tableNames);
}

return count($names) === count(array_intersect($names, $tableNames));

```
The current pull request allows to search for table names in different registers/cases - upper-camel-lower-title-mixed.

Reverse - compatible.
The function will give the correct result for any table names registers/cases - upper-camel-lower-title-mixed.

Thank you

### CODE REVIEW

**Comment by derrabus** [comment]
FTR: Continuation of #5922

File: src/Schema/AbstractSchemaManager.php [hunk 1]
```diff
@@ -309,7 +309,7 @@ protected function doListTableIndexes($table): array
      *
      * @throws Exception
      */
-    public function tablesExist($names)
+    public function tablesExist($names, bool $lowercase = true): bool
```
Comment (derrabus) [inline src/Schema/AbstractSchemaManager.php]: This is a breaking change.

We specifically allow userland code to create custom schema managers. Your change will break existing code that overrides `tableExists()`. You must not add new parameters or native return types to existing non-final methods.

**Review by derrabus — CHANGES_REQUESTED** [review]
Please don't open more and more pull requests for the same topic.

On your last PR I told you

> you haven't added a test that covers the problem you're trying to solve

And yet you still didn't care to add one. I really don't understand the problem that you want to solve here. A test could help me in that regard and it would make sure your use-case remains intact in the future.

Your PR description merely repeats the code you have written, but it does not explain the motivation behind your patch. 

Apart from that, static analysis uncovered a problem with your change that you need to address.

File: src/Schema/AbstractSchemaManager.php (hunk 1 above)
Comment (coresh) [reply src/Schema/AbstractSchemaManager.php]: Hi, @derrabus 
```
Re: "native return types to existing non-final methods":
```

[Done: https://github.com/doctrine/dbal/pull/5937/commits/627f7e9a6c01d165ff4f7ff930c048ae4bc6873b](https://github.com/doctrine/dbal/pull/5937/commits/627f7e9a6c01d165ff4f7ff930c048ae4bc6873b)

Thank you

Comment (coresh) [reply src/Schema/AbstractSchemaManager.php]:
```
Re: "Your PR description merely repeats the code you have written,
but it does not explain the motivation behind your patch."
```
https://github.com/doctrine/dbal/blob/3.6.x/src/Schema/AbstractSchemaManager.php -> `tablesExist()` -> Current implementation

allows to use the table name only using lowercase characters - lowercase table names.



### Details. Real world example.
Manually creating thousands of dynamically named tables
Case:
table: "users_**d**" -> already exists
table: "users_**D**" -> is NOT exists -> should be manually created
`users_D` - dynamically pre-generated name

Checking existed tables before creating required tables:
```
...
$tables = ["user_d", "user_D"];

$connection = $this->getEntityManager()->getConnection();

$schemaManager = $connection->getSchemaManager();

for ($i = 0; $i < count($tables); $i++) {

    $tableName = $tableNames[$i];

    if ($schemaManager->tablesExist([$tableName]) !== true) {

         $this->createAllDbTables->createTable($arguments, $tableName);
         $resultSet = $stmt->executeQuery();
         $result    = $resultSet->fetchAllAssociative();
         ...
    }
    ...
}
...

...
```

For tables:
```
$tableName = "user_d"; // "d" -> lowcased

$tableName = "user_D"; // "D" -> UPPERCASED
```
When called:
```
$schemaManager->tablesExist([$tableName])
```
`true` will be in both cases.

But, expected `false` for the table name "users_**D**"
 although the "users_**D**" table does not yet exist.

Patch - allowing to return the correct result for tables with a name in different case
by calling the method `tablesExist()`:
```
UPPERCASETABLENAME
CamelCaseTableName
lowercasetablename
Titlecasetablename
```
Thank you

**Comment by derrabus** [comment]
Okay, got it. Indeed, `tablesExist()` compares table names in a case-insensitive manner because that's how most DBMS behave. IIRC in MySQL and MariaDB this depends on the file system being used.

Long story short: If you want to write a portable application, you shouldn't have two tables with the same name that differ just by letter casing.

Is it actually a problem for you that table names are treated case-insensitively? If so, can you elaborate why?

**Comment by coresh** [comment]
> Okay, got it. Indeed, `tablesExist()` compares table names in a case-insensitive manner because that's how most DBMS behave. IIRC in MySQL and MariaDB this depends on the file system being used.
> 
> Long story short: If you want to write a portable application,
you shouldn't have two tables with the same
name that differ just by letter casing.
> 
> Is it actually a problem for you that table names are treated case-insensitively? If so, can you elaborate why?

Re:
```
Long story short: If you want to write a portable application, you shouldn't have two tables with the same name that differ just by letter casing.
```
Correct. Most projects used case-insensitive-naming for tables.
But, case-sensitive-naming, also possible and provided by the databases architecture.

Re:
```
most DBMS behave. IIRC in MySQL and MariaDB this depends on the file system being used.
```
Correct.
At the same time, databases make it possible to use the names of the names of tables in case-sensitively-mode.

Re:

```
Is it actually a problem for you that table names are treated case-insensitively? If so, can you elaborate why?
```
Correct. It's a problem.

E.g: Data stream is recorded in
tables, for example, the API provider provides data for a months with the name `M1` and for a minutes with the name `m1`, in order to eliminate the discrepancy, we have to:
add prefixes to table names first - e.g:
`unique_table_name_for_months_m1`
`unique_table_name_for_minutes_m1`,
at the same time, the use of prefixes can be avoided, e.g:
`M1`
`m1`

case-sensitively-mode-naming is very convenient when it comes to tens of thousands of tables

Thank you

**Comment by derrabus** [comment]
>> Long story short: If you want to write a portable application, you shouldn't have two tables with the same name that differ just by letter casing.
> 
> Correct. Most projects used case-insensitive-naming for tables. But, case-sensitive-naming, also possible and provided by the databases architecture.

Right, but that's not portable. If `m1` and `M1` and different tables in your app, this will not work on most SQL databases and even on MySQL it depends on the filesystem being used. And since we're building an abstraction layer and not a MySQL client library, I think it's okay for DBAL to exclude that case.

Accepting your change is a slippery slope because our whole schema management is currently built under the assumption that table names are treated case-insensitively by the DB. *If* we really wanted to go down that road, this won't be the only place we'd have to change.

This is mainly why I've decided to reject your change, the other reasons being that introducing a parameter is a breaking change and that your PR still lacks tests.

My advice for you would be to not rely on MySQL's case-sensitive treatment and instead use more distinct table names. If you really do need this behavior, I'd recommend overriding the schema manager in your app.

I thank you very much for your attempt to improve DBAL.

**Pull request closed without merging**